core/src/main/java/org/apache/iceberg/encryption/AesGcmOutputStream.java

```diff
@@ -0,0 +1,115 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.iceberg.encryption;
+
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.nio.ByteOrder;
+import org.apache.iceberg.io.PositionOutputStream;
+
+public class AesGcmOutputStream extends PositionOutputStream {
+
+  private final Ciphers.AesGcmEncryptor gcmEncryptor;
+  private final PositionOutputStream targetStream;
+  private final byte[] plainBlockBuffer;
+  private final byte[] fileAadPrefix;
+
+  private int positionInBuffer;
+  private long streamPosition;
+  private int currentBlockIndex;
+
+  AesGcmOutputStream(PositionOutputStream targetStream, byte[] aesKey, byte[] fileAadPrefix)
+      throws IOException {
+    this.targetStream = targetStream;
+    this.gcmEncryptor = new Ciphers.AesGcmEncryptor(aesKey);
+    this.plainBlockBuffer = new byte[Ciphers.PLAIN_BLOCK_SIZE];
+    this.positionInBuffer = 0;
+    this.streamPosition = 0;
```

Comment:
Shouldn't this always be `targetStream.pos() + positionInBuffer`?

Reply:
this value is returned by 

```java
@Override
  public long getPos() throws IOException {
    return streamPosition;
  }
```

  so is not related directly to targetStream.pos() and positionInBuffer.
I'll rename `positionInBuffer` to `positionInPlainBlock`, and `plainBlockBuffer` to `plainBlock`, to make the naming more consistent.

Reply:
I don't see why what I said wouldn't always be true. If so, we could eliminate this extra state.